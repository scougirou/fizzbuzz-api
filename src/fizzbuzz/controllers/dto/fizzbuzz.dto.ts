import { IsPositive } from 'class-validator';
import { ApiPropertyOptional } from '@nestjs/swagger';

export const MAXIMUM_LIMIT = 1000;
export const DEFAULT_LIMIT = 100;
export const MINIMUM_INTEGER = 0;
export const DEFAULT_INTEGER = 0;

export class FizzbuzzDto {
  @ApiPropertyOptional({
    description: 'First number to replace',
    minimum: MINIMUM_INTEGER,
    default: DEFAULT_INTEGER,
    example: 3,
  })
  @IsPositive()
  integer1: number;

  @ApiPropertyOptional({
    description: 'Second number to replace',
    minimum: MINIMUM_INTEGER,
    default: DEFAULT_INTEGER,
    example: 5,
  })
  @IsPositive()
  integer2: number;

  @ApiPropertyOptional({
    description:
      'Upper limit of the sequence count. The sequence will always start at 1.',
    minimum: MINIMUM_INTEGER,
    maximum: MAXIMUM_LIMIT,
    default: DEFAULT_LIMIT,
    example: 10,
  })
  @IsPositive()
  limit: number;

  @ApiPropertyOptional({
    description: 'First replacement string',
    example: 'fizz',
    default: 'empty string',
  })
  replacement1: string;

  @ApiPropertyOptional({
    description: 'Second replacement string',
    example: 'buzz',
    default: 'empty string',
  })
  replacement2: string;
}
