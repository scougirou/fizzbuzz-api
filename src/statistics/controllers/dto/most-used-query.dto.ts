import { FizzbuzzDto } from '../../../fizzbuzz/controllers/dto/fizzbuzz.dto';
import { IsDate, IsOptional, IsPositive } from 'class-validator';

export class MostUsedQueryDto {
  @IsOptional()
  query: FizzbuzzDto;
  @IsPositive()
  count: number;
  @IsDate()
  lastUsed: Date;
}
