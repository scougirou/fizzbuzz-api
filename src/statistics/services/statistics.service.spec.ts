import { Test, TestingModule } from '@nestjs/testing';
import { StatisticsService } from './statistics.service';
import { MostUsedQueryDto } from '../controllers/dto/most-used-query.dto';
import { FizzbuzzDto } from '../../fizzbuzz/controllers/dto/fizzbuzz.dto';
import { plainToClass } from 'class-transformer';

/**
 * Wrapper around the stat service for testing purpose.
 * Expose private members/methods that would not be otherwise accessible from Jest.
 */
class StatisticsServiceWrapper extends StatisticsService {
  constructor() {
    super();
  }
  async logUsage(input: FizzbuzzDto): Promise<void> {
    return super.logUsage(input);
  }

  getMostUsedQuery(): MostUsedQueryDto {
    return super.getMostUsedQuery();
  }

  setStats(stats: Map<string, number>): void {
    this.stats = stats;
  }
}

describe('StatisticsService', () => {
  let service: StatisticsServiceWrapper;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [StatisticsServiceWrapper],
    }).compile();

    service = module.get<StatisticsServiceWrapper>(StatisticsServiceWrapper);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should return the most used query', () => {
    const query1: FizzbuzzDto = plainToClass(FizzbuzzDto, {
      integer1: 3,
      integer2: 5,
      limit: 10,
      replacement1: 'fizz',
      replacement2: 'buzz',
    });
    const query2: FizzbuzzDto = plainToClass(FizzbuzzDto, {
      integer1: 0,
      integer2: 0,
      limit: 100,
      replacement1: 'foo',
      replacement2: 'bar',
    });
    const stats = new Map<string, number>([
      [JSON.stringify(query1), 10],
      [JSON.stringify(query2), 100],
    ]);
    service.setStats(stats);
    const expectedResult: MostUsedQueryDto = {
      query: query2,
      count: 100,
    };

    const result = service.getMostUsedQuery();

    expect(result).toStrictEqual(expectedResult);
  });

  it('should return the an empty object of no stats', () => {
    const result = service.getMostUsedQuery();

    expect(result).toBeUndefined();
  });
});
