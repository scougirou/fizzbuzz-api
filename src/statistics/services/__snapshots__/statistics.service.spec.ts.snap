// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatisticsService should add stats to the stat 1`] = `
Map {
  "{\\"integer1\\":3,\\"integer2\\":5,\\"limit\\":10,\\"replacement1\\":\\"fizz\\",\\"replacement2\\":\\"buzz\\"}" => Object {
    "count": 1,
    "lastUsed": 2021-10-18T14:52:19.212Z,
  },
}
`;

exports[`StatisticsService should should not log more than 10 stats 1`] = `
Map {
  "{\\"integer1\\":0,\\"integer2\\":5,\\"limit\\":10,\\"replacement1\\":\\"fizz\\",\\"replacement2\\":\\"buzz\\"}" => Object {
    "count": 1,
    "lastUsed": 2021-10-18T14:52:19.212Z,
  },
  "{\\"integer1\\":1,\\"integer2\\":5,\\"limit\\":10,\\"replacement1\\":\\"fizz\\",\\"replacement2\\":\\"buzz\\"}" => Object {
    "count": 1,
    "lastUsed": 2021-10-18T14:52:19.212Z,
  },
  "{\\"integer1\\":2,\\"integer2\\":5,\\"limit\\":10,\\"replacement1\\":\\"fizz\\",\\"replacement2\\":\\"buzz\\"}" => Object {
    "count": 1,
    "lastUsed": 2021-10-18T14:52:19.212Z,
  },
  "{\\"integer1\\":3,\\"integer2\\":5,\\"limit\\":10,\\"replacement1\\":\\"fizz\\",\\"replacement2\\":\\"buzz\\"}" => Object {
    "count": 1,
    "lastUsed": 2021-10-18T14:52:19.212Z,
  },
  "{\\"integer1\\":4,\\"integer2\\":5,\\"limit\\":10,\\"replacement1\\":\\"fizz\\",\\"replacement2\\":\\"buzz\\"}" => Object {
    "count": 1,
    "lastUsed": 2021-10-18T14:52:19.212Z,
  },
  "{\\"integer1\\":5,\\"integer2\\":5,\\"limit\\":10,\\"replacement1\\":\\"fizz\\",\\"replacement2\\":\\"buzz\\"}" => Object {
    "count": 1,
    "lastUsed": 2021-10-18T14:52:19.212Z,
  },
  "{\\"integer1\\":6,\\"integer2\\":5,\\"limit\\":10,\\"replacement1\\":\\"fizz\\",\\"replacement2\\":\\"buzz\\"}" => Object {
    "count": 1,
    "lastUsed": 2021-10-18T14:52:19.212Z,
  },
  "{\\"integer1\\":7,\\"integer2\\":5,\\"limit\\":10,\\"replacement1\\":\\"fizz\\",\\"replacement2\\":\\"buzz\\"}" => Object {
    "count": 1,
    "lastUsed": 2021-10-18T14:52:19.212Z,
  },
  "{\\"integer1\\":8,\\"integer2\\":5,\\"limit\\":10,\\"replacement1\\":\\"fizz\\",\\"replacement2\\":\\"buzz\\"}" => Object {
    "count": 1,
    "lastUsed": 2021-10-18T14:52:19.212Z,
  },
  "{\\"integer1\\":9,\\"integer2\\":5,\\"limit\\":10,\\"replacement1\\":\\"fizz\\",\\"replacement2\\":\\"buzz\\"}" => Object {
    "count": 1,
    "lastUsed": 2021-10-18T14:52:19.212Z,
  },
  "{\\"integer1\\":10,\\"integer2\\":5,\\"limit\\":10,\\"replacement1\\":\\"fizz\\",\\"replacement2\\":\\"buzz\\"}" => Object {
    "count": 1,
    "lastUsed": 2021-10-18T14:52:19.212Z,
  },
  "{\\"integer1\\":11,\\"integer2\\":5,\\"limit\\":10,\\"replacement1\\":\\"fizz\\",\\"replacement2\\":\\"buzz\\"}" => Object {
    "count": 1,
    "lastUsed": 2021-10-18T14:52:19.212Z,
  },
  "{\\"integer1\\":12,\\"integer2\\":5,\\"limit\\":10,\\"replacement1\\":\\"fizz\\",\\"replacement2\\":\\"buzz\\"}" => Object {
    "count": 1,
    "lastUsed": 2021-10-18T14:52:19.212Z,
  },
  "{\\"integer1\\":13,\\"integer2\\":5,\\"limit\\":10,\\"replacement1\\":\\"fizz\\",\\"replacement2\\":\\"buzz\\"}" => Object {
    "count": 1,
    "lastUsed": 2021-10-18T14:52:19.212Z,
  },
  "{\\"integer1\\":14,\\"integer2\\":5,\\"limit\\":10,\\"replacement1\\":\\"fizz\\",\\"replacement2\\":\\"buzz\\"}" => Object {
    "count": 1,
    "lastUsed": 2021-10-18T14:52:19.212Z,
  },
}
`;
